### PR TITLE
feat(next): make defineConfig generic

### PR DESCRIPTION
**Pull request by florian-lefebvre**
## Changes

- Makes `defineConfig` generic (backward compatible). That means that some parts of the config will be constrained by others. For example, it will ensure that `i18n.defaultLocale` is one of `i18n.locales` at the type level
- Reworks how `astro/config` works, by having one source of truth (from `config.d.ts`/`config.mjs` to `src/config/entrypoint.ts` being compiled)

## Testing

Should still pass, I added types tests

## Docs

I'll do a docs PR about i18n once approved

### CODE REVIEW

**Review by ematipico — APPROVED** [review]
I have no idea what the new TS code is doing 😅 but I saw the demo, and I liked it!

File: packages/astro/src/types/public/config.ts [hunk 1]
```diff
@@ -1205,30 +1215,31 @@ export interface AstroUserConfig {
...
-		 * The default locale of your website/application. This is a required field.
+		 * A list of all locales supported by the website. This is a required field.
 		 *
-		 * No particular language format or syntax is enforced, but we suggest using lower-case and hyphens as needed (e.g. "es", "pt-br") for greatest compatibility.
+		 * Languages can be listed either as individual codes (e.g. `['en', 'es', 'pt-br']`) or mapped to a shared `path` of codes (e.g.  `{ path: "english", codes: ["en", "en-US"]}`). These codes will be used to determine the URL structure of your deployed site.
+		 *
+		 * No particular language code format or syntax is enforced, but your project folders containing your content files must match exactly the `locales` items in the list. In the case of multiple `codes` pointing to a custom URL path prefix, store your content files in a folder with the same name as the `path` configured.
 		 */
-		defaultLocale: string;
+		locales: [TLocales] extends [never] ? Locales : TLocales;
```
Comment (florian-lefebvre) [inline packages/astro/src/types/public/config.ts:1229]: Since the generic uses `locales` as the source of truth, I've updated the reference so that a user first defines `locales`, then pick a `defaultLocale` based on the already defined locales. I'll submit a PR to the beta docs

File: .changeset/forty-trains-notice.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+'astro': minor
+---
+
+Makes `defineConfig` generic
```
Comment (ematipico) [inline .changeset/forty-trains-notice.md:5]: I'm not sure we should use this phrase. There isn't any generic that **the user should provide**. 

I would rephrase the changeset in a way that highlights the auto completion and the type safety.

File: .changeset/forty-trains-notice.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+
+TypeScript will now provide feedback at the type level when using `defineConfig`. For example, it will make sure `i18n.defaultLocale` is one of the locales declared in `i18n.locales`
```
Comment (florian-lefebvre) [inline .changeset/forty-trains-notice.md:7]:
```suggestion
Improves `defineConfig` type safety. TypeScript will now error if a group of related options do not have coherent types. For example, it will now error if `i18n.defaultLocale` is not one of the locales specified in `i18n.locales`
```

File: packages/astro/src/config/index.ts [hunk 4]
```diff
@@ -52,7 +61,15 @@ export function getViteConfig(
 					astroContentListenPlugin({ settings, logger, fs }),
 				],
 			},
-			{ settings, logger, mode, sync: false, manifest, ssrManifest: devSSRManifest },
+			{
+				settings,
+				logger,
+				// TODO: can the custom mode solve that?
```
Comment (florian-lefebvre) [inline packages/astro/src/config/index.ts:67]: Depends on #12150

File: packages/astro/test/types/define-config.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import { describe, it } from 'node:test';
+import { defineConfig } from '../../dist/config/index.js';
+import type { AstroUserDefineConfig } from '../../dist/types/public/config.js';
+import { expectTypeOf } from 'expect-type';
+
+describe('defineConfig()', () => {
```
Comment (ematipico) [inline packages/astro/test/types/define-config.ts:6]: Since both `fallback` and `domains` were changed too, I think we should add a test for them

**Review by ematipico — APPROVED** [review]

File: packages/astro/test/types/define-config.ts (hunk 5 above)
Comment (florian-lefebvre) [reply packages/astro/test/types/define-config.ts:6]: It doesn't matter here. The generic only depends on `i18n.locales`, and I specify `i18n.defaultLocale` because it's required

File: packages/astro/src/config/entrypoint.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+		config,
+	};
+}
+
+export function passthroughImageService(): ImageServiceConfig {
+	return {
+		entrypoint: 'astro/assets/services/noop',
+		config: {},
+	};
+}
```
Comment (bluwy) [inline packages/astro/src/config/entrypoint.ts:19]: I think these two are missing their jsdoc?

File: packages/astro/src/config/entrypoint.ts
Comment (bluwy) [inline packages/astro/src/config/entrypoint.ts:1]: I think we should also add a note here to keep this file as light as possible as it affects the Astro config loading time. It also looks like we're importing more than needed on startup now, unrelated to this PR, but we should revisit optimizing this later.

File: packages/astro/src/types/public/config.ts [hunk 7]
```diff
@@ -102,7 +110,9 @@ export interface ViteUserConfig extends OriginalViteUserConfig {
  * Astro User Config
  * Docs: https://docs.astro.build/reference/configuration-reference/
  */
-export interface AstroUserConfig {
+export type AstroUserConfig = AstroUserDefineConfig<never>;
+
+export interface AstroUserDefineConfig<TLocales extends Locales = never> {
```
Comment (bluwy) [inline packages/astro/src/types/public/config.ts:115]: Since this file is being re-exported from `types/index.ts`, the types for `AstroUserDefineConfig` will be public anyway, so maybe it's not worth separating this as two types? `defineConfig` with its generic already exposed it anyways.

If we want to completely hide this, we may need to update `types/index.ts` to only export the ones we need. But I'm not sure how to handle `defineConfig`.

Given that it's hard to handle this, I personally think it's easier to expose generic and call it a day. It's unlikely we need to break it again.

Comment (florian-lefebvre) [reply packages/astro/src/types/public/config.ts:115]: I think it's fine to export `AstroUserDefineConfig` and mark it as internal. However, I don't think we should make `AstroUserConfig` generic because we're likely to make breaking changes to generics, eg. by adding more

Comment (bluwy) [reply packages/astro/src/types/public/config.ts]: Adding more isn't a breaking change though? The generics should all have `=` fallbacks, like the current one, `= never`. And if there's a breaking change there, wouldn't `defineConfig` also inherit the breaking change?

Comment (florian-lefebvre) [reply packages/astro/src/types/public/config.ts]: Mmmh I guess you're right, adding types isn't breaking. But I think I still prefer to play it safe by having another type where we can do whatever we want. Changing the shape of a generic is not excluded before the next major imo

File: .changeset/forty-trains-notice.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+'astro': minor
+---
+
+Improves `defineConfig` type safety. TypeScript will now error if a group of related options do not have coherent types. For example, it will now error if `i18n.defaultLocale` is not one of the locales specified in `i18n.locales`
```
Comment (sarah11918) [inline .changeset/forty-trains-notice.md:5]:
```suggestion
Improves `defineConfig` type safety. TypeScript will now error if a group of related configuration options do not have consistent types. For example, you will now see an error if your language set for `i18n.defaultLocale` is not one of the supported locales specified in `i18n.locales`.
```

**Review by sarah11918 — COMMENTED** [review]
Changeset looks good, but just checking that I don't see any specific error messages in this PR.

I'm fine with changing the order of `locales` and `defaultLocale`, but I'm not sure what if any docs changes would be required for this PR, since this doesn't change the fact that we do already say that the default locale must be in your list of supported locales. Happy for you to double check that in the docs repo though!

File: packages/astro/src/types/public/config.ts (hunk 7 above)
Comment (bluwy) [reply packages/astro/src/types/public/config.ts]: I still don't understand why that would make it safer, it's already being exposed via `defineConfig()` so any breaking changes between it and `AstroUserDefineConfig` will be linked.

But if you prefer to keep the way it is now (with `@internal`), I won't block the PR. It's not a huge deal at the end of the day 👍
